light-client-detector/src/examine.rs

```diff
@@ -0,0 +1,163 @@
+use tendermint::crypto::default::Sha256;
+use tendermint_light_client::{
+    state::State,
+    store::{memory::MemoryStore, LightStore},
+    verifier::types::LightBlock,
+    verifier::types::Status,
+};
+
+use super::{error::Error, provider::Provider, trace::Trace};
+
+// FIXME: Allow the hasher to be configured
+type Hasher = Sha256;
+
+pub fn examine_conflicting_header_against_trace(
+    trace: &Trace,
+    target_block: &LightBlock,
+    source: &Provider,
+) -> Result<(Trace, LightBlock), Error> {
+    let trusted_block = trace.first();
+
+    if target_block.height() < trusted_block.height() {
+        return Err(Error::target_block_lower_than_trusted(
+            target_block.height(),
+            trusted_block.height(),
+        ));
+    }
+
+    let mut previously_verified_block: Option<LightBlock> = None;
+
+    for trace_block in trace.iter() {
+        let result = examine_conflicting_header_against_trace_block(
+            source,
+            trace_block,
+            target_block,
+            previously_verified_block,
+        )?;
+
+        match result {
+            ExaminationResult::Continue(prev_verified_block) => {
+                previously_verified_block = Some(prev_verified_block);
+                continue;
+            },
+            ExaminationResult::Divergence(source_trace, trace_block) => {
+                return Ok((source_trace, trace_block));
+            },
+        }
+    }
+
+    // We have reached the end of the trace. This should never happen. This can only happen if one of the stated
+    // prerequisites to this function were not met.
+    // Namely that either trace[len(trace)-1].Height < targetBlock.Height
+    // or that trace[i].Hash() != targetBlock.Hash()
+    Err(Error::no_divergence())
+}
+
+#[derive(Debug)]
+pub enum ExaminationResult {
+    Continue(LightBlock),
+    Divergence(Trace, LightBlock),
+}
+
+fn examine_conflicting_header_against_trace_block(
```

Comment:
I think we need a sync review for this function. I got lost in some of the places when reviewing from the IBC usecase where we have a trace with two blocks, first is trusted the second/last is the `target_block`.

Reply:
as discussed in the sync review, let's try to cleanup this function a bit so it reads easier.

Reply:
Had a go at it here: [f33040f](https://github.com/informalsystems/tendermint-rs/pull/1292/commits/f33040f9b6b1e8d9232d9821da32b965d6ad3ae7)